lib/src/atom/matcher.rs

```diff
@@ -578,20 +618,114 @@ impl From<&[(VariableAtom, Atom)]> for Bindings {
     fn from(pairs: &[(VariableAtom, Atom)]) -> Self {
         let mut bindings = Bindings::new();
         for (var, val) in pairs {
-            let mut result_vec = match val {
+            bindings = match val {
                 Atom::Variable(val) => bindings.add_var_equality(&var, &val),
                 _ => bindings.add_var_binding_v2(var, val),
-            };
-            match result_vec.len() {
-                0 => { bindings = Bindings::new(); },
-                1 => { bindings = result_vec.pop().unwrap(); },
-                _ => { panic!("Error: Multiple Bindings created from Atoms slice"); }
-            }
+            }.unwrap_or_else(|e| panic!("Error creating Bindings from Atoms: {}", e));
         }
         bindings
     }
 }
 
+/// Represents a set of [Bindings] instances resulting from an operation where multiple matches are possible.
+#[derive(Clone, Debug)]
+pub struct BindingsSet(smallvec::SmallVec<[Bindings; 1]>);
+
+impl core::iter::FromIterator<Bindings> for BindingsSet {
+    fn from_iter<I: IntoIterator<Item=Bindings>>(iter: I) -> Self {
+        let new_vec = iter.into_iter().collect();
+        BindingsSet(new_vec)
+    }
+}
+
+impl IntoIterator for BindingsSet {
+    type Item = Bindings;
+    type IntoIter = smallvec::IntoIter<[Bindings; 1]>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl From<Bindings> for BindingsSet {
+    fn from(bindings: Bindings) -> Self {
+        BindingsSet(smallvec::smallvec![bindings])
+    }
+}
+
+impl TryFrom<BindingsSet> for Bindings {
+    type Error = Atom;
+    fn try_from(mut set: BindingsSet) -> Result<Self, Atom> {
+        match set.len() {
+            0 => Ok(Bindings::new()),
+            1 => Ok(set.0.pop().unwrap()),
+            _ => Err(Atom::expr([ERROR_SYMBOL, sym!("Set Contains Multiple Bindings")]))
+        }
+    }
+}
+
+impl Extend<Bindings> for BindingsSet {
+    fn extend<I: IntoIterator<Item=Bindings>>(&mut self, iter: I) {
+        self.0.extend(iter);
+    }
+}
+
+impl core::ops::Deref for BindingsSet {
+    type Target = [Bindings];
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl BindingsSet {
+
+    pub fn empty() -> Self {
+        BindingsSet(smallvec::smallvec![])
+    }
+
+    pub fn new() -> Self {
+        BindingsSet(smallvec::smallvec![Bindings::new()])
+    }
+
+    /// An internal function to execute an operation that may take a single Bindings instance and replace
+    /// it with a new BindingsSet containing zero or more Bindings instances
+    fn perform_one_to_many_op<F>(mut self, mut func: F) -> Self
+        where F: FnMut(Bindings) -> BindingsSet
+    {
+        let mut new_set = BindingsSet::empty();
+        for bindings in self.0.drain(..) {
+            new_set.extend(func(bindings).into_iter());
```

Comment:
It does absolutely nothing because it's implied.  So I'll delete it if you think it's more readable without it.  It was a holdover from before I implemented the `IntoIterator` trait on `BindingsSet`